cirq-core/cirq/ops/boolean_hamiltonian.py

```diff
@@ -112,6 +114,171 @@ def _decompose_(self):
         )
 
 
+def _gray_code_comparator(k1: Tuple[int, ...], k2: Tuple[int, ...], flip: bool = False) -> int:
+    """Compares two Gray-encoded binary numbers.
+
+    Args:
+        k1: A tuple of ints, representing the bits that are one. For example, 6 would be (1, 2).
+        k2: The second number, represented similarly as k1.
+        flip: Whether to flip the comparison.
+
+    Returns:
+        -1 if k1 < k2 (or +1 if flip is true)
+        0 if k1 == k2
+        +1 if k1 > k2 (or -1 if flip is true)
+    """
+    max_1 = k1[-1] if k1 else -1
+    max_2 = k2[-1] if k2 else -1
+    if max_1 != max_2:
+        return -1 if (max_1 < max_2) ^ flip else 1
+    if max_1 == -1:
+        return 0
+    return _gray_code_comparator(k1[0:-1], k2[0:-1], not flip)
+
+
+def _simplify_commuting_cnots(
+    cnots: List[Tuple[int, int]], flip_control_and_target: bool
+) -> Tuple[bool, List[Tuple[int, int]]]:
+    """Attempts to commute CNOTs and remove cancelling pairs.
+
+    Commutation relations are based on 9 (flip_control_and_target=False) or 10
+    (flip_control_target=True) of [4]:
+    When flip_control_target=True:
+
+         CNOT(j, i) @ CNOT(j, k) = CNOT(j, k) @ CNOT(j, i)
+    ───X───────       ───────X───
+       │                     │
+    ───@───@───   =   ───@───@───
+           │             │
+    ───────X───       ───X───────
+
+    When flip_control_target=False:
+
+    CNOT(i, j) @ CNOT(k, j) = CNOT(k, j) @ CNOT(i, j)
+    ───@───────       ───────@───
+       │                     │
+    ───X───X───   =   ───X───X───
+           │             │
+    ───────@───       ───@───────
+
+    Args:
+        cnots: A list of CNOTS, encoded as integer tuples (control, target).
+        flip_control_and_target: Whether to flip control and target.
+
+    Returns:
+        A Boolean that tells whether a simplification has been performed.
+        The CNOT list, potentially simplified.
+    """
+
+    target, control = (0, 1) if flip_control_and_target else (1, 0)
+
+    i = 0
+    qubit_to_index: Dict[int, int] = {cnots[i][control]: i} if cnots else {}
+    for j in range(1, len(cnots)):
+        if cnots[i][target] != cnots[j][target]:
+            # The targets (resp. control) don't match, so we reset the search.
+            i = j
+            qubit_to_index = {cnots[j][control]: j}
+            continue
+
+        if cnots[j][control] in qubit_to_index:
+            k = qubit_to_index[cnots[j][control]]
+            # The controls (resp. targets) are the same, so we can simplify away.
+            cnots = [cnots[n] for n in range(len(cnots)) if n != j and n != k]
+            return True, cnots
```

Comment:
Why do we return immediately upon finding a cancellation? It seems like we could do this more efficiently by continuing the search, to avoid repeating the section we already searched.

Same for `_simplify_cnots_triplets`.

Reply:
So every time I simplify, the number of entries goes down and so my loop index might be invalid. I read the paper and they are kind of not clear about in which order to go through. I am not sure how to safely resume the loop. Leaving unresolved.

Reply:
Agreed that resuming is non-trivial. It definitely seems possible, though - could you add a TODO here and reference the relevant issue (i.e. `TODO(#issue): {...}`)?

Reply:
Created #4532 and added two TODOs.